aiosmtpd/smtp.py

```diff
@@ -213,18 +207,33 @@ def _handle_client(self):
                 return
             try:
                 # XXX this rstrip may not completely preserve old behavior.
-                line = line.decode('utf-8').rstrip('\r\n')
+                line = line.rstrip(b'\r\n')
                 log.info('%r Data: %s', self.session.peer, line)
                 if not line:
                     yield from self.push('500 Error: bad syntax')
                     continue
-                i = line.find(' ')
+                i = line.find(b' ')
+                # Decode to string only the command name part, which must be
+                # ASCII as per RFC.  If there is an argument, it is decoded to
+                # UTF-8/surrogateescape so that non-UTF-8 data can be
+                # re-encoded back to the original bytes when the SMTP command
+                # is handled.
                 if i < 0:
-                    command = line.upper()
+                    command = line.upper().decode(encoding='ascii')
                     arg = None
                 else:
-                    command = line[:i].upper()
+                    command = line[:i].upper().decode(encoding='ascii')
                     arg = line[i+1:].strip()
+                    # Remote SMTP servers can send us UTF-8 content despite
+                    # whether they've declared to do so or not.  Some old
+                    # servers can send 8-bit data.  Use surrogateescape so
+                    # that the fidelity of the decoding is preserved, and the
+                    # original bytes can be retrieved.
+                    if self.enable_SMTPUTF8:
+                        arg = str(
+                            arg, encoding='utf-8', errors='surrogateescape')
+                    else:
+                        arg = str(arg, encoding='ascii', errors='strict')
```

Comment:
This line of course can fail with a `UnicodeDecodeError`, and as you see in the `test_bad_encoded_param()` test, the exception will get logged because of the wide `except Exception as error` at the end of `_handle_client()`.  I think actually we should *not* log it, and we should provide a better 500 error message.  The reasoning is that it's not a bug in the implementation if `enable_SMTPUTF8` is False but non-ASCII is given by the client.  Another way of saying it is that we should expect that failure mode.

```
test_bad_encoded_param (aiosmtpd.tests.test_smtp.TestStrictASCII) ... ERROR:mail.log:SMTP session exception
Traceback (most recent call last):
  File "/home/barry/projects/aiosmtpd/aiosmtpd/smtp.py", line 236, in _handle_client
    arg = str(arg, encoding='ascii', errors='strict')
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 11: ordinal not in range(128)
ok
```

Perhaps we should add a special catch for this, and handle it a little differently, e.g.

```diff
2 files changed, 12 insertions(+), 2 deletions(-)
aiosmtpd/smtp.py            | 12 +++++++++++-
aiosmtpd/tests/test_smtp.py |  2 +-

modified   aiosmtpd/smtp.py
@@ -233,7 +233,17 @@ class SMTP(asyncio.StreamReaderProtocol):
                         arg = str(
                             arg, encoding='utf-8', errors='surrogateescape')
                     else:
-                        arg = str(arg, encoding='ascii', errors='strict')
+                        try:
+                            arg = str(arg, encoding='ascii', errors='strict')
+                        except UnicodeDecodeError:
+                            # This happens if enable_SMTPUTF8 is false, meaning
+                            # that the server explicitly does not want to
+                            # accept non-ASCII, but the client ignores that and
+                            # sends non-ASCII anyway.
+                            yield from self.push(
+                                '500 Error: strict ASCII mode')
+                            # Should we yield from self.handle_exception()?
+                            continue
                 max_sz = (self.command_size_limits[command]
                           if self.session.extended_smtp
                           else self.command_size_limit)
modified   aiosmtpd/tests/test_smtp.py
@@ -914,7 +914,7 @@ class TestStrictASCII(unittest.TestCase):
             client.send(b'MAIL FROM: <anne\xFF@example.com>\r\n')
             code, response = client.getreply()
             self.assertEqual(code, 500)
-            self.assertIn(b'Error: (UnicodeDecodeError)', response)
+            self.assertIn(b'Error: strict ASCII mode', response)
 
     def test_mail_param(self):
         with SMTP(*self.address) as client:
```

The same reasoning applies in `smtp_DATA()`, but I'll comment on that below.